changelogs/fragments/66451-win_dns_client-dhcp-support.yml

```diff
@@ -0,0 +1,7 @@
+minor_changes:
+  - Network adapter information is returned now, including adapter name, index, GUID, and nameserver information.
```

Comment:
This should not be done in the same module. A new module like `win_dns_client_info`, or some other `win_*_info` module should be used to report information. Keep this to making changes.

Reply:
Is there some guidance on this in the docs that I missed? I've seen some newer modules coming up that are only for gathering info, but wasn't aware we were to actively avoid returning info.

In any case will remove it; asking so I can read up more the design guidelines.

Reply:
I can't see anything that explicitly states don't return resource info in a normal module but it does say on https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html#contributing-to-ansible-objective-requirements the following about facts and info modules

> if a module is named <something>_facts, it should be because its main purpose is returning ansible_facts. Do not name modules that do not do this with _facts. Only use ansible_facts for information that is specific to the host machine, for example network interfaces and their configuration, which operating system and which programs are installed.

> Modules that query/return general information (and not ansible_facts) should be named _info. General information is non-host specific information, for example information on online/cloud services (you can access different accounts for the same online service from the same host), or information on VMs and containers accessible from the machine.

In this case we could see a benefit of it being _facts but personally I would recommend it being _info. I find _facts modules are good if retrieving a snapshot of the whole but if you wanted a info on a particular instance of that resource I think the _info model fits better.

But going back to why it should be in a separate module;

1. It adds more complexity to code that should be focused on changing the state idempotently
2. When dealing with check mode it's hard to properly return info, should it be of the state it would have changed to or the state it is
3. What if a user only wanted the info and not actually make a change, we would have to do something like `state: info` which isn't really a resource state but an action the user wants to take

The disadvantage of splitting it out into 2 separate modules is now we have code duplication. This can be somewhat negated by module utils but even without that I find the benefits of separate modules outweigh the negatives.

Finally you might find some instances of modules that do what I'm saying you shouldn't do (like win_service). These are largely historical aberrations and were done before some of these rules and guidelines have been put into place.

Reply:
Thanks for the feedback, much appreciated!

For 1. I feel like most of the information needed for return is already needed to make changes, so complexity and code-duplication is minimal. That's what happened here, I realized I had all this info and could return it fairly trivially. Writing the documentation was the most time consuming part!

I feel 2, for sure. I actually struggled with that in this and went with returning changed (final) state, for reasons that lead into 3).

For 3, the way I've done this with the aforementioned historical modules is to just do `check_mode: yes`. (`win_service` is an easier example because you can just give a name with no other params to get info). And that's also where I felt returning final state made sense; in check mode the final state is the current state, which felt intuitive to me but maybe isn't for everyone.

Other disadvantages of splitting into 2 separate modules imo is having to execute 2 separate modules if you want to modify and get information. Depending on what machine your controller is and over what connection you're running, individual task runs can be somewhat slow. If you want to make a change and then get information, calling 2 modules increases execution time.

In loops, the loop will also have to be duplicated, and depending on the situation, might not line up well (the same issue of trying to loop over a block and having to do a dynamic include of tasks to treat them as atomic).

To be clear, I don't think it makes sense to _not_ have `_info` and `_facts` modules; I actually love them. I've encountered enough modules that (AWS especially) that claimed to support check mode and then made changes and that's scary, so having dedicated modules to return info is useful in giving confidence to an end user that "this module doesn't change anything". But it's also nice to have info returned during a change. 

For code deduplication, `module_utils` of course makes sense. But also if the (changer) module returns current state in check mode well enough, then it would also be possible for the `_info` version of a module to become a minimal action plugin that just calls the full module with `check_mode` switched on, and have no module code needed at all.

Anyway, just thoughts and ramblings. Not trying to go against the tide and I will make the requested changed here! 

I do think it would be helpful to be explicit about that guidance in the relevant docs alongside the (already good) info about how to implement `_info`/`_facts` modules.